csrc/ops/composite.cpp

```diff
@@ -71,7 +73,20 @@ TensorView* matmul(TensorView* a, TensorView* b) {
   bcast_dims.at(bcast_dims.size() - 1) = false;
   bcast_dims.at(bcast_dims.size() - 3) = true;
   auto* tv1b = broadcast(b, bcast_dims);
-  return fusedMultiplySum(tv0b, tv1b, {-2});
+
+  NVF_CHECK(
+      a->getDataType().value() == b->getDataType().value(),
+      "data types of inputs to matmul don't match");
+  auto* output = fusedMultiplySum(tv0b, tv1b, {-2});
+  if (cast_output_to_input) {
+    // For matmul, the output dtype should match input.
+    return maybeCastOp(a->getDataType().value(), output);
+  }
+  return output;
+}
+
+TensorView* matmul(TensorView* a, TensorView* b) {
+  return matmul(a, b, true /* cast output to input dtype */);
```

Comment:
I think the static cast thing here might be easily changed... https://github.com/NVIDIA/Fuser/blob/866f10a186c697ab8e915455325085794899372d/csrc/python_frontend/python_bindings.cpp#L1140

But I don't know if serde `Binary_TV` for matmul might have more static casts down the line.

Reply:
As @jjsjann123 mentioned the problem came from creating an OpRecord where the function being called doesn't match the exact signature of the args. I think it can be solved, but I didn't think it was worth it at this point.